File: challenge8.go
/*
Alex Edwards
Nov, 2014
challenge8.go

This script creates a performance Cloud Server and assigns an A record to it
It will then create a monitoring check and alarm for the public IP
It then returns the IP Address, FQDN, and Monitoring Entity ID
*/

package main

import(
    "github.com/miguel-branco/goconfig"
    "github.com/rackspace/gophercloud"
    "github.com/rackspace/gophercloud/rackspace"
    "github.com/rackspace/gophercloud/rackspace/compute/v2/images"
    "github.com/rackspace/gophercloud/rackspace/compute/v2/flavors"
    "github.com/rackspace/gophercloud/rackspace/compute/v2/servers"
    "fmt"
    "log"
    "strings"
)

func main() {
    // Read in config
    c, conferr := goconfig.ReadConfigFile("./.rackspace_credentials")
    if conferr != nil {
        log.Fatalf("Failed to parse config data: %s", conferr)
    }

    // Define vars
    user, _ := c.GetString("rackspace", "username")
    apikey, _ := c.GetString("rackspace", "apikey")
    user = strings.Trim(user, "\"")
    apikey = strings.Trim(apikey, "\"")
    
    authopts := gophercloud.AuthOptions{
        Username: user,
        APIKey: apikey,
    }

    // Get our main provider
    provider, autherr := rackspace.AuthenticatedClient(authopts)
    if autherr != nil {
        log.Fatalf("Failed to authenticate: %s", autherr)
    }

    // Get server provider
    serviceClient, _ := rackspace.NewComputeV2(provider, gophercloud.EndpointOpts{
        Region: "ORD",
    })
    
    // Get image
    image, _ := images.Get(serviceClient, "ffa476b1-9b14-46bd-99a8-862d1d94eb7a").Extract()
    // Get Ubuntu 12 flavor
    flavor, _ := flavors.Get(serviceClient, "performance1-1").Extract()

    // Create our new server
    server, servercreateerr := servers.Create(serviceClient, servers.CreateOpts{
        Name:      "challenge8",
        ImageRef:  image.ID,
        FlavorRef: flavor.ID,
    }).Extract()

    if servercreateerr != nil {
        log.Fatalf("Failed to create server: %s", servercreateerr)
    }

    // Wait for server creation to finish
    fmt.Println("Server creation process started, waiting for it to complete.")
    serverpenderr := servers.WaitForStatus(serviceClient, server.ID, "ACTIVE", 600)

    if serverpenderr != nil {
        log.Fatalf("Failed to create server: %s", serverpenderr)
    }

}

File: challenge8.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Alex Edwards
# Nov, 2014
# challenge6.py
#
# This script creates a Cloud Performance 1G Server, and adds an A record for it.
# It will then create a cloud monitoring check and alarm based on ping of the public IP
# Finally it returns the public IP, FQDN, and Monitoring Check ID
#

from __future__ import print_function
import pyrax
import six
import sys
try:
    import configparser
except ImportError:
    import ConfigParser as configparser

# Setup our credentials
config_file="./.rackspace_credentials"
config = configparser.ConfigParser()
config.read(config_file)
username = config.get('rackspace', 'username').replace('"', '')
apikey = config.get('rackspace', 'apikey').replace('"', '')

pyrax.set_setting("identity_type", "rackspace")
pyrax.set_credentials(username, apikey)

# Create objects for managing servers, dns, monitoring
cs = pyrax.cloudservers
dns = pyrax.cloud_dns
cm = pyrax.cloud_monitoring

# Get flavor and image
flavor = [flavor for flavor in cs.flavors.list() if flavor.ram == 1024][0]
image = [img for img in cs.images.list() if "Ubuntu 12" in img.name][0]

# Create our server
server = cs.servers.create("challenge8", image.id, flavor.id)
print("Server creation process started")

# Wait for server
server = pyrax.utils.wait_until(server, "status", ["ACTIVE", "ERROR"])
print("Done!")

# Get domain to add A record for
dom = pyrax.cloud_dns.find(name='drunken-donuts.com')

# Add in two records, for www.@ and @
recs = [{
        "type": "A",
        "name": "drunken-donuts.com",
        "data": server.accessIPv4,
        "ttl": 3600,
        }, {
        "type": "A",
        "name": "www.drunken-donuts.com",
        "data": server.accessIPv4,
        "ttl": 3600,
        }]
records = dom.add_records(recs)

# Create entity for check
ent = cm.create_entity(name="entity_for_server", ip_addresses={"example": server.accessIPv4},
        metadata={"description": "I am Totoro"})

# Create our check
chk = cm.create_check(ent, label="ping_check", check_type="remote.ping",
        details={"url": server.accessIPv4}, period=900,
        timeout=20, monitoring_zones_poll=["mzdfw", "mzlon", "mzsyd"],
        target_hostname=server.accessIPv4)

# Create our notification plan should check fail
# I don't want to be emailed...no alert
np = cm.create_notification_plan(label="default")

# Create our alarm definitions to determine if check fails
alarm = cm.create_alarm(ent, chk, np, "if (rate(metric['average']) > 10) { return new AlarmStatus(WARNING); } return new AlarmStatus(OK);")

# Print out the details
print("IP Address: " + server.accessIPv4)
print("FQDN: " + records[1].name)
print("Check ID: " + chk.id)
